### ISSUE TITLE
Meson sometimes decides to write files into meson-out

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#6623]
For whatever reason meson sometimes decides that it should write files into the meson-out folder instead of the proper mirrored file structure, this can lead to problems when you generate multiple files of the same name.

See this [mesa bug](https://gitlab.freedesktop.org/mesa/mesa/issues/2512). I've also hit this with one of my projects, although it's not as serious for me because I don't have any code generators.

**Comment by dcbaker** [comment]
This appears to be related to meson changing the layout mode to flat for some reason. I need to investigate more.

**Comment by abergmeier** [comment]
Meson version is `0.51.0` if that helps.

**Comment by dcbaker** [comment]
I've been able to intermittently reproduce this on 0.52.1 as well

**Comment by DJSchaffner** [comment]
Happens in `0.53.1` aswell. Although i can't reproduce it consistently, it seems to happen more regularly when generating a new build folder.

**Comment by dcbaker** [comment]
Interesting, I've only been able to reproduce it by causing meson to reconfigure due to an upgrade.

**Comment by nirbheek** [comment]
Maybe this could happen if something fiddly happened in the coredata? We could merge something like this to make it barf if that happens:

```diff
--- a/mesonbuild/backend/backends.py
+++ b/mesonbuild/backend/backends.py
@@ -204,10 +204,13 @@ class Backend:

     @lru_cache(maxsize=None)
     def get_target_dir(self, target):
-        if self.environment.coredata.get_builtin_option('layout') == 'mirror':
+        layout = self.environment.coredata.get_builtin_option('layout')
+        if layout == 'mirror':
             dirname = target.get_subdir()
-        else:
+        elif layout == 'flat':
             dirname = 'meson-out'
+        else:
+            raise AssertionError('Invalid layout value {!r} found in coredata'.format(layout))
         return dirname

     def get_target_dir_relative_to(self, t, o):
```

**Comment by nloomans** [comment]
This issue also happens for me, multiple times a day. Meson randomly decides to put its files in `$BUILDDIR/meson-out` instead of `$BUILDDIR`.

Any files you would like me to collect the next time this happens?

I am using meson `0.53.1` on `macOS Mojave 10.14.6` installed using homebrew.

**Comment by nloomans** [comment]
@nirbheek I applied your patch to my local meson install. Let's see if it barks

**Comment by nloomans** [comment]
So, I did some more debugging and I found the root cause of my issue. I have an extension for vscode called [vscode-meson](https://github.com/asabil/vscode-meson) for which I enabled the `mesonbuild.configureOnOpen` option.

I now discovered that the extension configured meson with the `--layout=flat` option by default 😅. This can be resolved by either setting `mesonbuild.configureOnOpen` to false or changing `mesonbuild.configureOptions` so it doesn't set the layout to flat.
